src/Data/Text/Utf8.hs

```diff
@@ -4,16 +4,35 @@
 -- Licensed under the 3-clause BSD license, see the LICENSE file in the
 -- repository root.
 
+{-# LANGUAGE BangPatterns #-}
 {-# LANGUAGE MagicHash #-}
 
-module Data.Text.Utf8 (CodeUnit, CodeUnitIndex(..), Text(..), unpackUtf8, stringToByteArray, indexTextArray, unicode2utf8, pack) where
+module Data.Text.Utf8
+    ( CodeUnit
+    , CodeUnitIndex (..)
+    , Data.Text.Utf8.readFile
+    , Text (..)
+    , decode2
+    , decode3
+    , decode4
+    , indexTextArray
+    , pack
+    , stringToByteArray
+    , toLowerAscii
+    , unicode2utf8
+    , unpackUtf8
+    ) where
 
-import Data.Bits (Bits, shiftR, (.&.), (.|.))
+import Data.Bits (Bits (shiftL), shiftR, (.&.), (.|.))
 import Data.Primitive.ByteArray (ByteArray (ByteArray), byteArrayFromList, indexByteArray,
-                                 sizeofByteArray)
+                                 newByteArray, sizeofByteArray, unsafeFreezeByteArray,
+                                 writeByteArray)
 import Data.Word (Word8)
 
+import qualified Data.ByteString as BS
 import Data.Char (ord)
+import qualified Data.Char as Char
```

Comment:
nit: can you put all the qualified imports in a new block below the unqualified ones? That makes it consistent with the rest, and easy to see at a glance where identifiers come from, and what we depend on. The order we use is:

- unqualified external
- qualified external
- unqualified internal
- qualified internal

with an empty line between each block.